crates/natrix/src/reactivity/signal_list.rs

```diff
@@ -0,0 +1,563 @@
+    /// since their indices shift.
+    pub fn insert(&mut self, index: usize, value: T) {
+        if index > self.items.len() {
+            log_or_panic!("Index {} out of bounds for insert (len {})", index, self.items.len());
+            return;
+        }
+
+        self.items.insert(index, value);
+        
+        // Insert a new dep list at this position
+        if let Ok(mut deps) = self.deps.try_borrow_mut() {
+            deps.insert(index, SignalDepList::new());
+        } else {
+            log_or_panic!("Deps list already borrowed during insert");
+        }
+
+        // Invalidate readers from this index onwards (since indices shifted)
+        if let Ok(mut deps) = self.deps.try_borrow_mut() {
+            for dep_list in deps.iter_mut().skip(index) {
+                core::statics::reg_dirty_list(|| dep_list.create_iter_and_clear());
+            }
+        } else {
+            log_or_panic!("Deps list already borrowed during insert invalidation");
+        }
+        
+        core::statics::reg_dirty_list(|| {
+            self.len_deps.get_mut().create_iter_and_clear()
+        });
+    }
+
+    /// Remove an element at the specified index.
+    /// WARNING: This invalidates readers at this index and all subsequent indices.
+    pub fn remove(&mut self, index: usize) -> Option<T> {
+        if index >= self.items.len() {
+            log_or_panic!("Index {} out of bounds for remove (len {})", index, self.items.len());
+            return None;
+        }
+
+        let result = self.items.remove(index);
+        
+        // Remove the dep list for this index and invalidate it immediately
+        if let Ok(mut deps) = self.deps.try_borrow_mut() {
+            let mut removed_deps = deps.remove(index);
+            core::statics::reg_dirty_list(|| removed_deps.create_iter_and_clear());
+        } else {
+            log_or_panic!("Deps list already borrowed during remove");
+        }
+        
+        // Invalidate readers from this index onwards
+        if let Ok(mut deps) = self.deps.try_borrow_mut() {
+            for dep_list in deps.iter_mut().skip(index) {
+                core::statics::reg_dirty_list(|| dep_list.create_iter_and_clear());
+            }
+        } else {
+            log_or_panic!("Deps list already borrowed during remove invalidation");
+        }
+        
+        core::statics::reg_dirty_list(|| {
+            self.len_deps.get_mut().create_iter_and_clear()
+        });
+
+        Some(result)
+    }
+
+    /// Iterate over the list elements.
+    /// This does NOT track dependencies - use explicit `.get(i)` calls for reactive iteration.
+    #[inline]
+    pub fn iter(&self) -> impl Iterator<Item = &T> {
+        self.items.iter()
+    }
+
+    /// Iterate over the list elements mutably.
+    /// This does NOT track dependencies or invalidate readers.
+    #[inline]
+    pub fn iter_mut(&mut self) -> impl Iterator<Item = &mut T> {
+        self.items.iter_mut()
+    }
```

Comment:
We should not expose this, proper reactive iteration is on the roadmap (and this struct will likely become the basis for that)